source/server/config_validation/filesystem.cc

```diff
@@ -0,0 +1,156 @@
+
+NullInstance::NullInstance(Filesystem::Instance& impl) : impl_(impl) {}
+
+Filesystem::FilePtr NullInstance::createFile(const std::string& path) {
+  return std::make_unique<NullFile>(path, makeErrorMessage(path));
+}
+
+bool NullInstance::fileExists(const std::string&) { return false; }
+
+bool NullInstance::directoryExists(const std::string&) { return false; }
+
+ssize_t NullInstance::fileSize(const std::string&) { return -1; }
+
+std::string NullInstance::fileReadToEnd(const std::string& path) {
+  throw EnvoyException(makeErrorMessage(path));
+}
+
+Filesystem::PathSplitResult NullInstance::splitPathFromFilename(absl::string_view path) {
+  return impl_.splitPathFromFilename(path);
+}
+
+bool NullInstance::illegalPath(const std::string& path) { return impl_.illegalPath(path); }
+
+std::string NullInstance::makeErrorMessage(const absl::string_view path) const {
+  return absl::StrCat("null file system does not include file: ", path);
+}
+
+RestrictedInstance::RestrictedInstance(absl::string_view allowed_prefix, Filesystem::Instance& impl)
+    : prefix_(allowed_prefix), impl_(impl) {}
+
+Filesystem::FilePtr RestrictedInstance::createFile(const std::string& path) {
+  if (absl::StartsWith(path, prefix_)) {
+    return impl_.createFile(path);
+  }
+  return std::make_unique<NullFile>(path, makeErrorMessage(path));
+}
+
+bool RestrictedInstance::fileExists(const std::string& path) {
+  if (absl::StartsWith(path, prefix_)) {
+    return impl_.fileExists(path);
+  }
+  return false;
+}
+
+bool RestrictedInstance::directoryExists(const std::string& path) {
+  if (absl::StartsWith(path, prefix_)) {
+    return impl_.directoryExists(path);
+  }
+  return false;
+}
+
+ssize_t RestrictedInstance::fileSize(const std::string& path) {
+  if (absl::StartsWith(path, prefix_)) {
+    return impl_.fileSize(path);
+  }
+  return -1;
+}
+
+std::string RestrictedInstance::fileReadToEnd(const std::string& path) {
+  if (absl::StartsWith(path, prefix_)) {
+    return impl_.fileReadToEnd(path);
+  }
+  throw EnvoyException(makeErrorMessage(path));
+}
+
+Filesystem::PathSplitResult RestrictedInstance::splitPathFromFilename(absl::string_view path) {
+  return impl_.splitPathFromFilename(path);
+}
+
+bool RestrictedInstance::illegalPath(const std::string& path) { return impl_.illegalPath(path); }
+
+std::string RestrictedInstance::makeErrorMessage(const absl::string_view path) const {
+  return absl::StrCat("allowed prefix ", prefix_, " does not include file: ", path);
+}
+
+InstancePtr makeValidationFilesystem(const Server::Options& options, Instance& real_file_system) {
+  if (options.mode() == Server::Mode::Load) {
+    if (options.configPath().empty()) {
+      return std::make_unique<NullInstance>(real_file_system);
+    }
+    return std::make_unique<RestrictedInstance>(options.configPath(), real_file_system);
+  }
+  return std::make_unique<WrapInstance>(real_file_system);
```

Comment:
You could return a nullptr here, and then call `makeValidationFilesystem()` in `validateConfig()`, check the result and then pass an appropriate `Filesystem::Instance&` the `ValidationInstance` ctor.

That would get rid of the boilerplate `WrapInstance`, but add some ugliness to `validateConfig()`. *I* think it would be a bit more maintainable, but I don't feel strongly about this.